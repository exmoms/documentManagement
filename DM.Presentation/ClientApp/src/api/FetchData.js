/**
 * @file this file includes functions that fetch data from API
 */

// Generic Delete Method
export async function deleteFromApi(url) {
  let res = await fetch(url, {
    method: "DELETE",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
  });
  return res;
}

// Get MetaData Models
export async function fetchData(url) {
  let object = {};
  /*
   * send the request through a CORS proxy
   * https://stackoverflow.com/questions/43262121/trying-to-use-fetch-and-pass-in-mode-no-cors/43268098#43268098
   */
  await fetch(url, {
    method: "GET",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
  })
    .then((response) => response.json())
    .then(
      (data) => (object = data),
      (error) => console.log(error)
    );
  return object;
}
