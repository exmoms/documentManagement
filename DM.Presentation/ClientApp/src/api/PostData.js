/**
 * @file This file includes functions that help to post data to a specific API
 * @version 1.0
 */

/**
 * Post new document data.
 * Upload form-data, scanned copy of doc and attachments.
 * @author Youssef Shaaban <y.shaaban@lit-co.net>
 * @param {*} documentData
 * @param {*} images
 * @param {*} attachments
 * @param {*} url
 * @param {*} option its values either "Add" for add new document version or "Update" to update exsiting document Version
 */
export async function postDocumentData(
  documentData,
  scannedPages,
  attachments,
  url = "",
  option
) {
  var formData = new FormData();

  if (process.env.NODE_ENV === "development") {
    console.log(documentData);
    console.log(attachments);
    console.log(scannedPages);
  }

  switch (option) {
    case "Add":
      formData.append("Document", JSON.stringify(documentData));
      break;
    case "Update":
      formData.append("Version", JSON.stringify(documentData));
      break;
    default:
  }

  attachments.forEach((attachment) => {
    formData.append("Attachments", attachment);
  });
  scannedPages.forEach((scannedPage) => {
    formData.append("Scans", scannedPage);
  });

  let response = {};
  await fetch(url, {
    // content-type header should not be specified!
    method: "POST",
    headers: {
      Accept: "application/json",
      // "Content-Type": "application/json",
    },
    body: formData,
  }).then(
    (res) => {
      response = res;
      // response.json();
    },
    (error) => console.log("[ERROR]:", error)
  );
  return response;
}

/**
 * Post Attachment Data.
 * Upload attachment json data + attachment file  .
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @param {*} payload
 * @param {*} attachmentFile
 * @param {*} url
 */
export async function postAttachmentData(payload, attachmentFile, url = "") {
  var formData = new FormData();

  formData.append("Attachment", JSON.stringify(payload));
  formData.append("FileAttachment", attachmentFile);

  let response = {};
  await fetch(url, {
    // content-type header should not be specified!
    method: "POST",
    body: formData,
  }).then(
    (res) => {
      response = res;
      // response.json();
    },
    (error) => console.log("[ERROR]:", error)
  );
  return response;
}

/**
 * Post Request To API
 * Send post request to api with  payload  and return a response.
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @param {*} url The api url
 * @param {*} payload js object which will be sent as request body.
 */
export async function postDataToAPI(url, payload = null, method = "POST") {

  const response = await fetch(url, {
    method: method,
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });
  return response;
}
