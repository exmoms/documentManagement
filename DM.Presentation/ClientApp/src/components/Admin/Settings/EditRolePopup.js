/**
 * @file Contains EditRolePopup component implementaion
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @version 1.0
 */

import {
  Button,
  FormControl,
  Input,
  InputLabel,
  MenuItem,
  Select,
} from "@material-ui/core";

import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogTitle from "@material-ui/core/DialogTitle";
import { withStyles } from "@material-ui/core/styles";
import React, { Component } from "react";
import { withTranslation } from "react-i18next";
import { postDataToAPI } from "../../../api/PostData";
import { Notification } from "../Notifications";

const ITEM_HEIGHT = 48;
const ITEM_PADDING_TOP = 8;
const MenuProps = {
  PaperProps: {
    style: {
      maxHeight: ITEM_HEIGHT * 4.5 + ITEM_PADDING_TOP,
      width: 400,
    },
  },
};

const styles = (theme) => ({
  paper: {
    maxWidth: "auto",
    marginTop: theme.spacing(4),
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
    justifyContent: "center",
  },
  margin: {
    margin: theme.spacing(2),
  },
});

class EditRolePopup extends Component {
  constructor(props) {
    super(props);
    this.state = {
      userId: this.props.user.userId,
      selectedRole: "",
      currentRole: "",
      availableRoles: ["Admin", "User"],
      error: false,
      errorMessage: [],
      showSuccess: false,
    };
    this.updateParentState = this.props.action;
  }
  t = this.props["t"];

  componentDidUpdate(prevProps) {
    if (
      prevProps.user.userId !== this.props.user.userId ||
      JSON.stringify(prevProps.user.roles) !==
        JSON.stringify(this.props.user.roles)
    ) {
      this.setState({
        userId: this.props.user.userId,
        selectedRole: this.props.user.roles[0],
        currentRole: this.props.user.roles[0],
      });
    }
  }

  toggleShow = (show, flag) => {
    this.updateParentState(show, flag);
  };

  handleChange = (event) => {
    this.setState({ selectedRole: event.target.value });
  };

  handelSubmit = async (event) => {
    event.preventDefault();

    const formData = {
      userId: this.props.user.userId,
      AssignedRoles: [],
      UnassignedRoles: [],
    };

    if (this.state.currentRole !== this.state.selectedRole) {
      formData.AssignedRoles.push(this.state.selectedRole);
      formData.UnassignedRoles.push(this.state.currentRole);
    }

    let response = await postDataToAPI("/api/user/EditUserRoles", formData);
    if (response.ok) {
      this.setState({ showSuccess: true });
    } else {
      let json = await response.json();
      this.setState({
        showSuccess: true,
        error: true,
        errorMessage: json.errorMessage,
      });
    }
    this.toggleShow(false, response.ok);
  };

  render() {
    const { classes } = this.props;
    let lang = localStorage.getItem("i18nextLng");
    return (
      <div>
        <Dialog
          position="centerCenter"
          open={this.props.show}
          onClose={() => this.toggleShow(false, false)}
          aria-labelledby="form-dialog-title"
          className={classes.paper}
        >
          <div
            style={
              lang === "ar" ? { textAlign: "right" } : { textAlign: "left" }
            }
          >
            <DialogTitle id="form-dialog-title">
              {this.t("edit_user_role")} : {this.props.user.userName}
            </DialogTitle>
          </div>

          <DialogContent>
            <div className={classes.paper}>
              <FormControl style={{ minWidth: 400, minHeight: 120 }}>
                <InputLabel id="demo-mutiple-chip-label">
                  {this.t("role_label")}
                </InputLabel>
                <Select
                  labelId="demo-mutiple-chip-label"
                  id="demo-mutiple-chip"
                  autoWidth
                  value={this.state.selectedRole}
                  onChange={this.handleChange}
                  input={<Input id="select-multiple-chip" />}
                  MenuProps={MenuProps}
                >
                  {this.state.availableRoles.map((name, index) => (
                    <MenuItem key={index} value={name}>
                      {name}
                    </MenuItem>
                  ))}
                </Select>
              </FormControl>
            </div>
          </DialogContent>
          <DialogActions>
            <Button
              id="submit-btn"
              onClick={this.handelSubmit}
              color="primary"
              className={classes.submit}
            >
              {this.t("submit_changes")}
            </Button>
            <Button
              id="CancelButton"
              onClick={() => this.toggleShow(false, false)}
              color="primary"
            >
              {this.t("button_cancel")}
            </Button>
          </DialogActions>
        </Dialog>
        <Notification
          id="notification"
          open={this.state.showSuccess}
          setOpen={(show) => this.setState({ showSuccess: show })}
          error={!!this.state.error}
          errorMessage={this.state.errorMessage}
        />
      </div>
    );
  }
}

export default withStyles(styles)(withTranslation()(EditRolePopup));
