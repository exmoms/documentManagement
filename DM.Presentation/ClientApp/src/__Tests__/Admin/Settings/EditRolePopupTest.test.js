import React from "react";
import ReactDom from "react-dom";
import EditRolePopup from "../../../components/Admin/Settings/EditRolePopup";
import { cleanup } from "@testing-library/react";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";
import TextField from "@material-ui/core/TextField";
import { act } from "react-dom/test-utils";
import Button from "@material-ui/core/Button";
import { BrowserRouter as Router } from "react-router-dom";
import * as FetchMock from "../../../api/FetchData";
import * as PostDataMock from "../../../api/PostData";
import Typography from "@material-ui/core/Typography";
import Dialog from "@material-ui/core/Dialog";
import DialogTitle from "@material-ui/core/DialogTitle";
import {
    FormControl,
    Input,
    InputLabel,
    MenuItem,
    Select,
  } from "@material-ui/core";
afterEach(cleanup);
describe("EditProfilePopup", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  test("renders without crashing", () => {
    const div = document.createElement("div");
    let user={userId:1 , userName:"userName A" , roles:["Admin"]};
    let action = jest.fn();
    let show= true;
    ReactDom.render(
        <EditRolePopup user={user} show={show} action={action}/>,
      div
    );
  });

  test("matches snapshot", () => {
    let user={userId:1 , userName:"userName A" , roles:["Admin"]};
    let action = jest.fn();
    let show= true;
    const wrapper = shallow(
        <EditRolePopup user={user} show={show} action={action} />
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
  test("intialization EditRolePopup , onClose ,onChange", () => {
    let user={userId:1 , userName:"userName A" , roles:["Admin"]};
    let user2={userId:1 , userName:"userName B" , roles:["SuperAdmin"]};
    let action = jest.fn();
    let show= true;
  let  wrapper = mount(<EditRolePopup user={user} show={show} action={action}/>) ;
  expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
  expect(wrapper.find(DialogTitle).at(0).text()).toBe("edit_user_role : userName A");
  expect(wrapper.find("#demo-mutiple-chip").at(0).props().value).toBe("");
  expect(wrapper.find("#notification").at(0).props().open).toBeFalsy();
  expect(wrapper.find("#notification").at(0).props().error).toBeFalsy();
  expect(wrapper.find("#notification").at(0).props().errorMessage).toStrictEqual([]);
  act(()=>{
    wrapper.setProps({user:user2});
  })
  
  wrapper.render();
  wrapper.update();
  expect(wrapper.find(Dialog).at(0).props().open).toBeTruthy();
  expect(wrapper.find(DialogTitle).at(0).text()).toBe("edit_user_role : userName B");
  expect(wrapper.find("#demo-mutiple-chip").at(0).props().value).toBe("SuperAdmin");
  expect(wrapper.find("#notification").at(0).props().open).toBeFalsy();
  expect(wrapper.find("#notification").at(0).props().error).toBeFalsy();
  expect(wrapper.find("#notification").at(0).props().errorMessage).toStrictEqual([]);
  act(()=>{
    wrapper.find(Dialog).at(0).props().onClose();
  })
  
  wrapper.render();
  wrapper.update();
  expect(action).toHaveBeenCalledTimes(1);
 expect(action.mock.calls[0][0]).toBeFalsy();
 expect(action.mock.calls[0][1]).toBeFalsy();
 act(()=>{
    wrapper.find("#demo-mutiple-chip").at(0).props().onChange({ target: { value: 'SuperAdmins'} });
  })
  
  wrapper.render();
  wrapper.update();
  expect(wrapper.find("#demo-mutiple-chip").at(0).props().value).toBe("SuperAdmins");
});
test("handelSubmit when  res is ok and this.state.currentRole !== this.state.selectedRole is false", () => {
    const mockSuccessResponsePost ={ok :true};
    const mockJsonPromisePost = Promise.resolve(mockSuccessResponsePost);
    const PostMock = jest.spyOn(PostDataMock, "postDataToAPI").mockImplementation(() => mockJsonPromisePost);
    const preventDefaultMock = jest.fn();
    let user={userId:1 , userName:"userName A" , roles:["Admin"]};
    let user2={userId:1 , userName:"userName B" , roles:["SuperAdmin"]};
    let action = jest.fn();
    let show= true;
    let  wrapper = mount(<EditRolePopup user={user} show={show} action={action}/>);
    act(()=>{
        wrapper.setProps({user:user2});
      })
      wrapper.render();
      wrapper.update();
    return mockJsonPromisePost.then(() => {
        act(() => {
            wrapper
              .find("#submit-btn")
              .at(0)
              .props()
              .onClick({ preventDefault:preventDefaultMock});
          });
    }).then(() => {
        wrapper.render();
        wrapper.update();
        
        expect(preventDefaultMock).toHaveBeenCalledTimes(1);
        expect(PostMock).toHaveBeenCalledTimes(1);
        expect(PostMock.mock.calls[0][0]).toBe("/api/user/EditUserRoles");
        expect(PostMock.mock.calls[0][1]).toStrictEqual({"AssignedRoles": [], "UnassignedRoles": [], "userId": 1});
        expect(wrapper.find("#notification").at(0).props().open).toBeTruthy();
    });

  });

//   test("handelSubmit when  res isn' ok ", (done) => {
//     let mockSuccessErrorPromise = { error: ["errorMessage"] };
//     let mockSuccessError = Promise.resolve(mockSuccessErrorPromise);
//     const mockSuccessResponsePost = Promise.resolve({
//       ok: false,
//       json: () => mockSuccessError,
//     });
//     const PostMock = jest
//       .spyOn(PostDataMock, "postDataToAPI")
//       .mockReturnValue(mockSuccessResponsePost);
//     const preventDefaultMock = jest.fn();
//     let user={userId:1 , userName:"userName A" , roles:["user"]};
//     let user2={userId:1 , userName:"userName B" , roles:["Admin"]};
//     let action = jest.fn();
//     let show= true;
//     let wrapper = shallow(<EditRolePopup user={user} show={show} action={action}/>);
//     act(()=>{
//         wrapper.setProps({user:user2});
//       })
//       wrapper.render();
//       wrapper.update();
//     act(()=>{
//         wrapper.find("#demo-mutiple-chip").at(0).props().onChange({ target: { value: 'SuperAdmin'} });
//       })
//       wrapper.render();
//       wrapper.update();
//       expect(wrapper.find("#demo-mutiple-chip").at(0).props().value).toBe("SuperAdmin");
//     act(() => {
//         wrapper
//           .find("#submit-btn")
//           .at(0)
//           .props()
//           .onClick({ preventDefault:preventDefaultMock});
//       });
//     wrapper.update();
//     wrapper.render();

//     expect(preventDefaultMock).toHaveBeenCalledTimes(1);
//         expect(PostMock).toHaveBeenCalledTimes(1);
//         expect(PostMock.mock.calls[0][0]).toBe("/api/user/EditUserRoles");
//     expect(PostMock.mock.calls[0][1]).toStrictEqual();
//     process.nextTick(() => {

//       expect(wrapper.find("#notification").props().open).toBeTruthy();
//       expect(wrapper.find("#notification").props().error).toBeTruthy();
//       expect(wrapper.find("#notification").props().errorMessage).toStrictEqual([
//         "errorMessage",
//       ]);
//       expect(action).toHaveBeenCalledTimes(1);
//         expect(action.mock.calls[0][0]).toBeFalsy();
//       done(); 
//     });
//   });
});