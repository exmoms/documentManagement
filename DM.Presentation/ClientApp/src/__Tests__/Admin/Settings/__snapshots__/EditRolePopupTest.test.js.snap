// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditProfilePopup matches snapshot 1`] = `
<withI18nextTranslation(EditRolePopup)
  action={[MockFunction]}
  classes={
    Object {
      "margin": "withI18nextTranslation(EditRolePopup)-margin-2",
      "paper": "withI18nextTranslation(EditRolePopup)-paper-1",
    }
  }
  show={true}
  user={
    Object {
      "roles": Array [
        "Admin",
      ],
      "userId": 1,
      "userName": "userName A",
    }
  }
/>
`;
